core/src/banking_stage/transaction_scheduler/scheduler_controller.rs

```diff
@@ -181,8 +196,112 @@ impl SchedulerController {
+        let mut max_time_reached = false;
+        while !self.container.is_empty() {
+            let mut filter_array = [true; CHUNK_SIZE];
+            let mut ids = Vec::with_capacity(CHUNK_SIZE);
+            let mut txs = Vec::with_capacity(CHUNK_SIZE);
+
+            for _ in 0..CHUNK_SIZE {
+                if let Some(id) = self.container.pop() {
+                    ids.push(id);
+                } else {
+                    break;
+                }
+            }
+            let chunk_size = ids.len();
+            ids.iter().for_each(|id| {
+                let transaction = self.container.get_transaction_ttl(&id.id).unwrap();
+                txs.push(&transaction.transaction);
+            });
+
+            // use same filter we use for processing transactions:
+            // age, already processed, fee-check.
+            Self::pre_graph_filter(
+                &txs,
+                &mut filter_array,
+                &bank,
+                MAX_PROCESSING_AGE
+                    .saturating_sub(FORWARD_TRANSACTIONS_TO_LEADER_AT_SLOT_OFFSET as usize),
+            );
+
+            for (id, filter_result) in ids.iter().zip(&filter_array[..chunk_size]) {
+                if !*filter_result {
+                    self.container.remove_by_id(&id.id);
+                    continue;
+                }
+
+                ids_to_add_back.push(*id); // add back to the queue at end
+                let state = self.container.get_mut_transaction_state(&id.id).unwrap();
+                let sanitized_transaction = &state.transaction_ttl().transaction;
+                let immutable_packet = state.packet().clone();
+
+                // If not already forwarded and can be forwarded, add to forwardable packets.
+                if state.should_forward()
+                    && forwardable_packets.try_add_packet(
+                        sanitized_transaction,
+                        immutable_packet,
+                        feature_set,
+                    )
+                {
+                    saturating_add_assign!(num_forwarded, 1);
+                    state.mark_forwarded();
+                }
+            }
+
+            if start.elapsed() >= MAX_FORWARDING_DURATION {
+                max_time_reached = true;
+                break;
+            }
+        }
+
+        // Forward each batch of transactions
+        for batch in forwardable_packets.iter_batches() {
+            let _ = self.forwarder.forward_packets(
+                &ForwardOption::ForwardTransaction,
+                batch.get_forwardable_packets(),
+            );
+        }
+
+        // If we hit the time limit. Drop everything that was not checked/processed.
+        // If we cannot run these simple checks in time, then we cannot run them during
+        // leader slot.
+        if max_time_reached {
+            while let Some(id) = self.container.pop() {
+                self.container.remove_by_id(&id.id);
+            }
```

Comment:
seems like we could stand to implement a `clear()` on `TransactionStateContainer` and call `.clear()` on each of its members there

Reply:
it'd be similar, but not what we want. clearing would remove transaction that are "in-flight".

Example:

We are leader, schedule a transaction for processing.
We stop being leader, but processing thread hasn't yet sent the finished message back to scheduler thread.
If scheduler clears the container then we will remove the id from the map, and when scheduler gets the finished message then it's for an unknown id (since if we cleared it).